### PR TITLE
Ensure that register typing constraints are respected at join points

### PR DESCRIPTION
**Pull request by mshinwell**
It has been known for some time that great care has to be taken when using Obj.magic to ensure that, at join points in the control flow graph, a register "known to hold only immediates" does not end up being used to hold a value that might be a pointer.  If this happens, it will not be registered as a GC root, which can cause it to become invalid.  (The typical fix is something like using None instead of "unit".)

After having fixed a piece of code having that problem this morning it became apparent that the following function, which does not use Obj.magic, suffers from the same problem:

```
type 'a typ = Int : int typ | Ptr : int list typ

let f (type a) (t : a typ) (p : int list) : a =
  match t with
  | Int -> 100
  | Ptr -> p
```

If the function "f" is inlined into another function then the situation above can arise, as exhibited in the test case on this pull request, which segfaults.

The only reasonable solution it would appear at present is to ensure that at join points we respect the register typing.  This may produce slightly worse code in some cases.  For such cases the correct solution to solve this problem is probably the propagation of more information about what values might be held in which identifiers right through from the type checker.

I think this is probably serious enough to warrant putting in 4.03.

### CODE REVIEW

File: asmcomp/cmm.ml [hunk 1]
```diff
@@ -32,6 +32,39 @@ let size_component = function
...
+  | Val, Int -> Val
+  | Val, Addr -> Val
+  | Val, Val -> Val
+  | Float, Float -> Float
+  | (Int | Addr | Val), Float
+  | Float, (Int | Addr | Val) ->
+    (* Float unboxing code must be sure to avoid this case. *)
+    assert false
+
+let ge_component comp1 comp2 =
```
Comment (ghost) [inline asmcomp/cmm.ml:69]: Couldn't this be implemented as `comp1 = lub_component comp1 comp2`? That would avoid having to check the code twice

Comment (mshinwell) [inline asmcomp/cmm.ml:69]: I tend to think it's clearer written out, actually.

**Comment by lpw25** [comment]
Patch looks correct to me.

**Comment by damiendoligez** [comment]
Devil's advocate here: the bug has been here since we introduced GADTs in 4.00.0. Is it really urgent to fix it now? What's your assessment of the risk level for this change?

**Comment by gasche** [comment]
My understanding is that the issue is sensibly more severe in presence of flambda than it was in previous releases, as it will aggressively inline these cases.

**Comment by mshinwell** [comment]
I'm also kind of surprised this hasn't caused trouble before, given how long the GADT support has been merged for.

It doesn't really seem an ideal change just before a release.  On balance I think I'm pretty convinced about the soundness of the change, but I'm not completely sure about potential performance degradation.  @xavierleroy : do you have an opinion on that?

**Comment by gasche** [comment]
Can you characterize the problematic joint points? Are they exactly those that match on a GADT at the source level? If so, that would make it easier to reason about potential performance impact (most of the code doesn't use GADTs and would be unaffected, and I understand there are a few performance-sensitive uses of GADTs that could be inspected manually).

Have you tried tracing the cases where the join decreases precision, for example on a compiler bootstrap?

**Comment by mshinwell** [comment]
I'm not sure I have an exact characterisation of the problematic join points, but I think a GADT match must be involved, with one of the branches returning a result that forces the backend to allocate a register of type "Int" for that result (in fact one of type "Addr" would probably cause it to go wrong too) and another branch returning a value that might be boxed.  In some cases the backend would then use the "Int" register as the result for the whole match, which might then go wrong if the live range for that register then extends forwards across a GC point (and if the boxed branch is taken in the match).

For performance, it's actually cases that do not use GADTs, but might end up getting penalised, which we're worried about.  One way we might be able to assess this is using OCamlPro's benchmarking suite which we were using for flambda, perhaps.

**Comment by mshinwell** [comment]
@garrigue Do you have an opinion on this one?

**Comment by lpw25** [comment]
> Patch looks correct to me.

Actually one thing I am not sure about: Is Addr supposed to include Int? At the moment if one branch returns Addr and another returns Int then they will both be stored in an Addr register. Is that correct? If Addr is allowed to hold Int then it is not clear to me what the difference between Val and Addr is.

**Comment by mshinwell** [comment]
Val is registered with the GC, Addr is not, as far as I recall.

**Comment by mshinwell** [comment]
(I will check about having Int in Addr)

**Comment by mshinwell** [comment]
Hmm, so I think if we use Addr occasionally rather than Int, it isn't that much of a big deal.  It will inhibit CSE at certain points.  That might be all, in fact, with the current code.

**Comment by bluddy** [comment]
Not to hijack the discussion, but could someone explain to me what the join points are?

**Comment by mshinwell** [comment]
When two control flow paths come together, such as at the end of a conditional expression.

**Comment by bluddy** [comment]
Thanks. And normally join points are places where we may register GC roots?

**Comment by mshinwell** [comment]
No.  The roots are effectively "registered" at a point where one OCaml function may call another.  The problem is that at the join point, when generating code, a decision has to be made about which register to use to hold the result of the conditional.  In some cases the backend reuses the result register from one of the two code paths (and generates a move at the end of the other code path into that register).  However this is wrong if the reused register will not be registered as a root (at any subsequent point where it might need to be, before it becomes dead) and the value being moved into it on the other code path may need to be registered.

**Comment by bluddy** [comment]
Awesome. Thanks again.

File: asmcomp/cmm.ml [hunk 2]
```diff
@@ -32,6 +32,39 @@ let size_component = function
...
   | Float -> Arch.size_float
 
+let lub_component comp1 comp2 =
+  match comp1, comp2 with
+  | Int, Int -> Int
+  | Int, Addr -> Addr
+  | Int, Val -> Val
+  | Addr, Int -> Addr
+  | Addr, Addr -> Addr
+  | Addr, Val -> Val
```
Comment (xavierleroy) [inline asmcomp/cmm.ml]: "Addr" denotes a derived pointer that might point in the middle of a heap block.  The GC is unable to treat such derived pointers correctly: it cannot follow them and it cannot update them after moving the underlying object.  This is why it is a compile-time fatal error to have a variable of type Addr live across a GC or call point.

Now, if you have a conditional where x : Addr in one branch and x : Val in the other, x may be a derived pointer and therefore must be given type Addr after the join point.  If it dies before the next GC / call point, everything will be fine.  Otherwise it is a serious code generation error and the compiler should bomb.

In summary: Addr, Val -> Addr and symmetrically.

Comment (mshinwell) [inline asmcomp/cmm.ml]: Ah yes, my mistake, the relation is indeed the other way around.

**Comment by mshinwell** [comment]
Patch updated following @xavierleroy 's comment

**Comment by xavierleroy** [comment]
Apart the Addr corner case, the patch looks good to me.  I'm not worried at all about performance degradation: most C-- variables, and especially most of those involved in a join point, are of type Val already.

**Comment by xavierleroy** [comment]
One last, cosmetic suggestion.  While checking the code I drew the lattice of types on a sheet of paper.  What about including it in a comment as ASCII art?  E.g.

```
      Addr     Float
       |
      Val
       |
      Int
```

**Comment by mshinwell** [comment]
Good idea.  I've added to the comment a note about [Addr] as well.

File: testsuite/tests/asmcomp/register_typing.ml [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+let f (type a) (t : a typ) (p : int list) : a =
+  match t with
+  | Int -> 100
+  | Ptr -> p
+
+let allocate_garbage () =
+  for i = 0 to 100 do
+    ignore (Array.make 200 0.0)
+  done
```
Comment (stedolan) [inline testsuite/tests/asmcomp/register_typing.ml:11]: Whether or not this function allocates anything on the minor heap is dependent on word size.

On 32-bit machines, a 200-float array is 400 words, bigger than Max_young_wosize, and nothing is allocated on the minor heap. Is this the intent?

File: asmcomp/selectgen.ml [hunk 4]
```diff
@@ -117,14 +117,19 @@ let join opt_r1 seq1 opt_r2 seq2 =
...
-        end else if Reg.anonymous r2.(i) then begin
+        end else if Reg.anonymous r2.(i)
+          && Cmm.ge_component r2.(i).typ r1.(i).typ
+        then begin
           r.(i) <- r2.(i);
           seq1#insert_move r1.(i) r2.(i)
         end else begin
-          r.(i) <- Reg.create r1.(i).typ;
+          let typ = Cmm.lub_component r1.(i).typ r2.(i).typ in
+          r.(i) <- Reg.create typ;
```
Comment (stedolan) [inline asmcomp/selectgen.ml:132]: The conditions here are somewhat subtle. I would feel safer if `#insert_move` was updated to add an assertion about `Cmm.ge_component`.

**Comment by stedolan** [comment]
Even in theory, what is the possible performance degradation of this patch? That the frame tables may be slightly larger?

**Comment by lpw25** [comment]
> Even in theory, what is the possible performance degradation of this patch?

Additional roots for the GC to try to mark even though they are always integers.

File: asmcomp/selectgen.ml (hunk 4 above)
Comment (mshinwell) [inline asmcomp/selectgen.ml:132]: Yeah, actually, that may be a good idea.  I think I thought of doing that some time ago in fact.  However I'm not sure if we should do that for 4.03 or just afterwards.  (I tend to think just afterwards.)

**Comment by xavierleroy** [comment]
Yes, larger frame tables, and a bit more time spent in the GC scanning the stack.
